proposals/20240606-jwt-handling.md

```diff
@@ -0,0 +1,135 @@
+## Scope
+This proposal aims to standardise how the Kubeflow backends should be handling the user information (name of user, groups they belong to), living in JWTs and http headers.
+
+This proposal takes as a requirement that users should be able to use K8s Tokens as `Authorization: Bearer <token>` headers in their request from inside the cluster. Note that the issuer of the tokens should not be relevant to the Kubeflow applications. It'll be up to the service-mesh to verify them, and be able to work with multiple issuers (i.e. Dex, K8s etc). The applications will only care about the user/groups information from the tokens.
+### In
+- Which component should be validating JWTs (id-tokens from OIDC or K8s ServiceAccount tokens)
+- Define where the backends should expect to find user related information
+- Define how different token issuers (i.e. Dex, K8s etc) should be handled
+### Out
+- Proposing code changes to existing components
+## Current State
+As of Kubeflow 1.8 the user information has been injected into requests as the `kubeflow-userid` header, from the AuthService (replaced by `oauth2-proxy`). For this approach to be secure there are the following patterns that Kubeflow follows:
+1. Backends in the `kubeflow` namespace that need to know the user identity rely on `kubeflow-userid` headers in http requests.
+2. The AuthService adds the `kubeflow-userid` header to all authenticated requests.
+3. Only requests from the Istio IngressGateway are trusted to have the `kubeflow-userid` header
+    1. Backends that are not exposed to user namespaces (i.e. jupyter-web-app) are only reachable via the Istio IngressGateway.
+    2. The KFP backend [explicitly drops requests](https://github.com/kubeflow/manifests/blob/96ce068e16b2a707464471bddc0d2a58e403d1fc/apps/pipeline/upstream/base/installs/multi-user/istio-authorization-config.yaml#L37) from user namespaces if they have this header
+    3. In-cluster Pods that want to talk to `kubeflow` workloads, which understands identity, are [using a K8s ServiceAccount Token](https://www.kubeflow.org/docs/components/pipelines/v1/sdk/connect-api/#full-kubeflow-subfrom-inside-clustersub)
+### Limitations
+- Not able to express `AuthorizationPolicies` for group header in Istio
+- Limited possibility to use custom JWT claims as a source of information about authenticated user
+
+To accommodate the above limitations and improve the authnz flow in terms of security, maintenance and flexibility
+we propose to add the JWT to the `Authorization` header, so it can be digested by Istio and have the user details securely
+injected to auth headers when authorized. This will also enable us to define policies in the future for better handling of groups.
+https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/
+
+But the above creates the following topics that require agreement on how to handle:
```

Comment:
require an agreement on how to handle them:

Reply:
pushed a commit a4ba96a4